### PR TITLE
BKNDLSS-29123 Multiple Select

### PR DESCRIPTION
**Pull request by Saharij**
Screenshot 2022-07-27 at 21 20 26
Screenshot 2022-07-27 at 21 17 17
Screenshot 2022-07-27 at 21 22 42

### CODE REVIEW

**Review by StasivPavlo — CHANGES_REQUESTED** [review]

File: components/bl-multiple-select-component/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/dist
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/.gitignore:1]: Could be removed

File: components/bl-multiple-select-component/src/helpers/use-multiple-select-class-list.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { useMemo } from 'react';
+
+export const useMultipleSelectClassList = props => {
+  const { disable, variant, classList } = props;
+
+  return useMemo(() => {
+    const classes = ['bl-customComponent-multipleSelect', variant, ...classList];
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/helpers/use-multiple-select-class-list.js:7]: use `cn` instead this hook

File: components/bl-multiple-select-component/src/helpers/validate.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+export const validate = items => {
+  if (!items) {
+    return [];
+  }
+  
+  if (typeof items === 'string') {
+    return items.split(',').map(item => ({
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/helpers/validate.js:7]: I think there also should be `.trim`

File: components/bl-multiple-select-component/src/multiple-select-field/chip.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export const Chip = ({ label }) => {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/chip.js:1]:
```
export const Chip = ({ label }) => (
  <div className="chip-item">
    <span className="chip-item__label">{ label }</span>
  </div>
)
```

File: components/bl-multiple-select-component/src/multiple-select-field/drop-down-button.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+export const DropDownButton = ({ isOptionsOpen }) => {
+  const dropDownButtonClasses = () => {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/drop-down-button.js:2]: use `cn`

File: components/bl-multiple-select-component/src/multiple-select-field/fieldset.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+export const Fieldset = ({ placeholder, multipleSelectValue, isMultipleSelectActive }) => {
+  const legendClasses = () => {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/fieldset.js:2]: use `cn`

File: components/bl-multiple-select-component/src/multiple-select-field/label.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+export const Label = props => {
+  const { multipleSelectId, placeholder, multipleSelectValue, isMultipleSelectActive } = props;
+  
+  const classes = () => {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/label.js:4]: use `cn`

File: components/bl-multiple-select-component/src/options/option.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { Checkmark } from './checkmark';
+
+export const Option = props => {
+  const { option, isOptionSelected, typeOfMultipleSelect, handleMultipleSelectValue } = props;
+
+  const optionClassList = () => {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/option.js:6]: use `cn`

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-multiple-select-component/src/multiple-select-field/input.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        className="input__container">
+        { typeOfMultipleSelect === "chip" ?
+          <div className="chip__items">
+            { multipleSelectValue.map(({ label, objectId }) => (
+              <Chip key={ objectId } label={ label } />
+            )) }
+          </div>
+          :
+          <span>{ inputValue }</span>
+        }
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/input.js:19]:
```suggestion
{ typeOfMultipleSelect === "chip" 
  ? <ChipList items={multipleSelectValue}/>
  : <span>{ inputValue }</span>
}
```

File: components/bl-multiple-select-component/src/multiple-select-field/input.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import { Chip } from './chip';
+
+export function Input(props) {
+  const { multipleSelectId, multipleSelectValue, typeOfMultipleSelect } = props;
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/input.js:4]: What about remove 'multiple' prefix? We know from context about select is multiple, and we have no other selects.

File: components/bl-multiple-select-component/component.json [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+          "value": "filled",
+          "label": "Filled"
+        }
+      ],
+      "showInSettings": true,
+      "hasLogicHandler": false
+    },
+    {
+      "type": "select",
+      "name": "typeOfMultipleSelect",
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/component.json:61]: selectType or type

File: components/bl-multiple-select-component/src/helpers/validate.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  if (!items) {
+    return [];
+  }
+  
+  if (typeof items === 'string') {
+    return items.trim().split(',').map(item => ({
+      objectId: item.trim(),
+      value: item.trim(),
+      label: item.trim(),
+    }));
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/helpers/validate.js:11]:
```suggestion
  return items
    .split(',')
    .map(i => i.trim())
    .map(i => ({ objectId: i, value: i, label: i }))
```

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-multiple-select-component/src/helpers/validate.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
+export const validate = items => {
+  if (!items) {
+    return [];
+  }
+
+  if (typeof items === 'string') {
+    return items.split(',').map(i => i.trim()).map(i => ({ objectId: i, value: i, label: i }));
+  }
+
+  if (typeof items === 'array') {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/helpers/validate.js:10]: image

Array.isArray

File: components/bl-multiple-select-component/src/helpers/validate.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+  return items;
+};
+
+const getId = () => {
+  const chr4 = () => Math.random().toString(16).slice(-4);
+  const chr8 = () => `${chr4()}${chr4()}`;
+
+  return `${chr8()}${chr8()}`;
+};
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/helpers/validate.js:29]:
```
const getId = () => Math.random().toString(16).slice(-8);
```

File: components/bl-multiple-select-component/src/index.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  useOnClickOutside(rootRef, handleClickOutside);
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div
+      ref={ rootRef }
+      className={ cn('bl-customComponent-multipleSelect', variant, ...classList, { disable }) }>
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/index.js:40]: I'm not sure, but `...classList` could be without `...`

File: components/bl-multiple-select-component/src/multiple-select-field/index.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
+import { useMemo } from "react";
+
+import { Label } from "./label";
+import { Input } from "./input";
+import { Fieldset } from "./fieldset";
+import { DropDownButton } from "./drop-down-button";
+
+export function SelectField(props) {
+  const {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/index.js:9]: in one line?

File: components/bl-multiple-select-component/src/options/index.js [hunk 17]
```diff
@@ -0,0 +1,64 @@
+import { useState, useRef, useEffect } from 'react';
+
+import { Option } from './option';
+
+export function Options(props) {
+  const {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/index.js:6]: in one line?

File: components/bl-multiple-select-component/src/options/index.js [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    <div className="options" style={{ transform: `translateY(-${ margin }px)` }} ref={ selectRef }>
+      { options.map(option => {
+        const isOptionSelected = selectValue.find(({ objectId }) => objectId === option.objectId);
+
+        return (
+          <Option
+            key={ option.objectId }
+            type={ type }
+            option={ option }
+            isOptionSelected={ !!isOptionSelected }
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/index.js:57]: !! should be not there - move it to place where you create variable

File: components/bl-multiple-select-component/src/options/index.js [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+    if (selectBottom > viewPortHeight) {
+      setMargin(selectBottom - viewPortHeight);
+    }
+  }, [selectRef])
+
+  const handleSelectValue = option => {
+    let newSelectValue;
+    const isOptionSelected = selectValue.find(({ objectId }) => objectId === option.objectId);
+
+    if (!isOptionSelected) {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/index.js:30]: swap conditions, positive should be first

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-multiple-select-component/src/multiple-select-field/label.js [hunk 20]
```diff
@@ -0,0 +1,13 @@
+const { cn } = BackendlessUI.CSSUtils;
+
+export function Label(props) {
+  const { selectId, placeholder, selectValue, isSelectActive } = props;
+
+  return (
+    <label
+      htmlFor={ selectId }
+      className={ cn("label", { ["label__move-up"]: isSelectActive || selectValue.length > 0 }) }>
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/label.js:9]: `selectValue.length > 0` => `selectValue.length`

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-multiple-select-component/README.md [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+<dd>A drop down list that displays all the options that you specify in the `Options` propertie.</dd>
+</dl>
+
+### Component Properties
+
+  Name         | Type       | Default value       | Description
+ --------------|------------|---------------------|-------------------------------------------------------------------------------------
+  Disable      | bool       | false               | This property allows you to disable component.
+  Placeholder  | string     | 'Placeholder'       | This property allows you to select a label of component.
+  Options      | { label: 'string', value: 'string' } |           | This property allows you to add items to be displayed in the drop down list. It's displayed on the component's logic page.
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/README.md:24]: What if some certain string values we describe as `'any string'`, but it just a typeof string we write `<string>`?

File: components/bl-multiple-select-component/src/multiple-select-field/index.js [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+    </div>
+  );
+};
+
+const getId = () => {
+  const chr4 = () => Math.random().toString(16).slice(-4);
+  const chr8 = () => `${chr4()}${chr4()}`;
+
+  return `${chr8()}${chr8()}`;
+};
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/multiple-select-field/index.js:40]:
```suggestion
const getId = () => Math.random().toString(16).slice(-8)
```

File: components/bl-multiple-select-component/src/options/index.js [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+  const [margin, setMargin] = useState(0);
+
+  useEffect(() => {
+    const viewPortHeight = window.innerHeight;
+    const selectBottom = selectRef.current?.getBoundingClientRect()?.bottom;
+
+    if (selectBottom > viewPortHeight) {
+      setMargin(selectBottom - viewPortHeight);
+    }
+  }, [selectRef])
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/index.js:18]: useMemo?

File: components/bl-multiple-select-component/src/multiple-select-field/index.js (hunk 22 above)
Comment (v-excelsior) [reply components/bl-multiple-select-component/src/multiple-select-field/index.js:40]: We have two similar getId functions, remove one of them

File: components/bl-multiple-select-component/src/index.js [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+
+  const rootRef = useRef(null);
+  const [isOptionsOpen, setIsOptionsOpen]= useState(false);
+  const [selectValue, setSelectValue] = useState([]);
+  const [isSelectActive, setIsSelectActive] = useState(false);
+
+  const optionsList = useMemo(() => validate(options), [options]);
+
+  const handleClickOutside = useCallback(() => {
+    if (!isOptionsOpen) {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/index.js:21]: positive condition first

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-multiple-select-component/README.md [hunk 25]
```diff
@@ -0,0 +1,33 @@
+# Multiple Select
```
Comment (Valodya) [inline components/bl-multiple-select-component/README.md:1]: please fix the readme to our standart

**Comment by Valodya** [comment]
please add a preview and thumbnail image

File: components/bl-multiple-select-component/src/helpers/use-on-click-outside.js [hunk 26]
```diff
@@ -0,0 +1,21 @@
+import { useEffect } from 'react';
+
+export const useOnClickOutside = (ref, handler) => {
```
Comment (Valodya) [inline components/bl-multiple-select-component/src/helpers/use-on-click-outside.js:3]:
```
export const useOnClickOutside = (ref, handler) => {
  useEffect(() => {
    const listener = event => {
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }
  
      handler(event);
    };

    document.addEventListener('mousedown', listener);
    document.addEventListener('touchstart', listener);

    return () => {
      document.removeEventListener('mousedown', listener);
      document.removeEventListener('touchstart', listener);
    };
  }, [ref, handler]);
};
```

File: components/bl-multiple-select-component/src/helpers/validate.js [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+
+  if (Array.isArray(items)) {
+    if (!items[0].hasOwnProperty('objectId')) {
+      return items.map(item => ({
+        ...item,
+        objectId: getId(),
+      }));
+    }
+
+    return items;
```
Comment (Valodya) [inline components/bl-multiple-select-component/src/helpers/validate.js:18]: this line is redundant

File: components/bl-multiple-select-component/src/multiple-select-field/input.js [hunk 28]
```diff
@@ -0,0 +1,22 @@
+import { ChipList } from './chip-list';
+
+export function Input(props) {
```
Comment (Valodya) [inline components/bl-multiple-select-component/src/multiple-select-field/input.js:3]:
```
export function Input({ type, selectId, selectValue }) {
```

File: components/bl-multiple-select-component/src/options/index.js [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    setSelectValue(newSelectValue);
+
+    if (onChange) {
+      onChange({ selectValue: newSelectValue });
+    }
+  };
+
+  return (
```
Comment (Valodya) [inline components/bl-multiple-select-component/src/options/index.js:41]:
```
const selectedValuesMap = useMemo(() => {
  return selectValue.reduce((m, { objectId }) => ({...m, [objectId]:1 }), {})
},[selectValue])

return (
    <div className="options" style={{ transform: `translateY(-${ margin }px)` }} ref={ selectRef }>
      { options.map(option => (
          <Option
            key={ option.objectId }
            type={ type }
            option={ option }
            isOptionSelected={ selectedValuesMap[option.objectId] }
            handleSelectValue={ handleSelectValue }
          />
        )) }
    </div>
  );
```

File: components/bl-multiple-select-component/src/options/option.js [hunk 30]
```diff
@@ -0,0 +1,16 @@
+import { Checkmark } from './checkmark';
+
+const { cn } = BackendlessUI.CSSUtils;
+
+export function Option(props) {
```
Comment (Valodya) [inline components/bl-multiple-select-component/src/options/option.js:5]:
```
export function Option({ type, option, selected, onSelect }) {
```

File: components/bl-multiple-select-component/src/multiple-select-field/label.js [hunk 31]
```diff
@@ -0,0 +1,13 @@
+const { cn } = BackendlessUI.CSSUtils;
+
+export function Label(props) {
+  const { selectId, placeholder, selectValue, isSelectActive } = props;
+
+  return (
+    <label
+      htmlFor={ selectId }
+      className={ cn("label", { ["label__move-up"]: isSelectActive || selectValue.length }) }>
```
Comment (Valodya) [inline components/bl-multiple-select-component/src/multiple-select-field/label.js:9]: `cn("label", { 'label__move-up': isSelectActive || selectValue.length })`

File: components/bl-multiple-select-component/src/options/option.js [hunk 32]
```diff
@@ -0,0 +1,16 @@
...
+  const { type, option, isOptionSelected, handleSelectValue } = props;
+
+  return (
+    <div
+      onClick={ () => handleSelectValue(option) }
+      className={ cn("option", { ['option__selected']: isOptionSelected }) }>
```
Comment (Valodya) [inline components/bl-multiple-select-component/src/options/option.js:11]: `cn('option', { option__selected: isOptionSelected })`

**Pull request merged**